<a href="https://colab.research.google.com/github/Pritija11/workshop-1/blob/main/wrkshp_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd
def cost_function(X, Y, W):

  m = len(Y)
  cost = np.sum((X.dot(W)-Y) **2)/(2*m)
  return cost



In [3]:
def gradient_descent(X, Y, W, alpha, iterations):
  cost_history = [0] * iterations
  m = len(Y)
  for iteration in range(iterations):
    Y_pred = X.dot(W)
    loss = Y_pred - Y
    dw = (X.T.dot(loss))/(m)
    W_update = W - alpha * dw
    cost = cost_function(X, Y, W_update)
    cost_history[iteration] = cost
  return W_update, cost_history


In [4]:
def rmse(Y, Y_pred):

  rmse = np.sqrt(np.mean((Y - Y_pred) ** 2))
  return rmse


In [5]:
def r2(Y, Y_pred):
  mean_y = np.mean(Y)
  ss_tot = np.sum((Y - mean_y) ** 2)
  ss_res = np.sum((Y - Y_pred) ** 2)
  r2 = 1 - (ss_res / ss_tot)
  return r2


In [7]:
from sklearn.model_selection import train_test_split

def main():
  data = pd.read_csv("/content/drive/MyDrive/dataset/student.csv")
  X = data[['Math', 'Reading']].values
  Y = data['Writing'].values
  X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

  W = np.zeros(X_train.shape[1])
  alpha = 0.001
  iterations = 1000

  W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)

  Y_pred = np.dot(X_test, W_optimal)

  model_rmse = rmse(Y_test, Y_pred)
  model_r2 = r2(Y_test, Y_pred)

  print("Final Weights:", W_optimal)
  print("Cost History (First 10 iterations):", cost_history[:10])
  print("RMSE on Test Set:", model_rmse)
  print("R-Squared on Test Set:", model_r2)

if __name__ == "__main__":
      main()



Final Weights: [4.7978325  5.02019875]
Cost History (First 10 iterations): [np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246)]
RMSE on Test Set: 615.8660062668006
R-Squared on Test Set: -1514.236975362205


1. Did your Model Overfitt, Underfitts, or performance is acceptable.

When alpha = 0.00001, The model is underfitting as the RMSE value is very high and R-square is negative. The cost history is slao constant which states Gradient descent did not converge properly.

2. Experiment with different value of learning rate, making it higher and lower, observe the result.

alpha = 0.0001:
Final Weights: [0.47978325 0.50201988]
Cost History (First 10 iterations): [np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211)]
RMSE on Test Set: 6.092665443799174
R-Squared on Test Set: 0.851706281452247

When alpha value is increased from 0.00001 to 0.0001, the model perfomance improved. The RMSE reduced from 63.36 to 6.09, and the R² score increased from −15.04 to 0.85, indicating that the model learned effectively. So, the model perfomance is acceptable neither underfitting nor overfitting.

aplha = 0.001:
Final Weights: [4.7978325  5.02019875]
Cost History (First 10 iterations): [np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246)]
RMSE on Test Set: 615.8660062668006
R-Squared on Test Set: -1514.236975362205

The model is unstable and performs poorly.

